Rpackage/R/mmif.cv.R

```diff
@@ -0,0 +1,109 @@
+#' @param scorer The Loss calculation function 
+#' @param n_cpu The number of cores to register for parallel programing of the code, default value is 1 and n_cpu = -1 to select all cores.
+#' 
+#' @return The list consist of best score, best parameters and list of all parameter values with cross validation score . 
+#' 
+#' @author Toby Dylan Hocking, Alexandre Drouin, Torsten Hothorn, Parismita Das
+#' 
+#' @examples
+#' library(mmit)
+#' target.mat <- rbind(
+#'   c(0,1), c(0,1), c(0,1),
+#'   c(2,3), c(2,3), c(2,3))
+#' 
+#' feature.mat <- rbind(
+#'   c(1,0,0), c(1,1,0), c(1,2,0),
+#'   c(1,3,0), c(1,4,0), c(1,5,0))
+#' 
+#' colnames(feature.mat) <- c("a", "b", "c")
+#' feature.mat <- data.frame(feature.mat)
+#' 
+#' param_grid <- NULL
+#' param_grid$max_depth <- c(Inf, 4, 3)
+#' param_grid$margin <- c(2, 3, 5)
+#' param_grid$min_sample <- c(2, 5, 10)
+#' param_grid$loss <- c("hinge")
+#' param_grid$n_trees <- c(10, 20, 30)
+#' param_grid$n_features <- c(as.integer(ncol(feature.mat)**0.5))
+#' 
+#' result <- mmit.cv(target.mat, feature.mat, param_grid, scorer = mse, n_cpu = -1)
+#' 
+#' @export
+mmif.cv <- structure(function(target.mat, feature.mat, 
+                              param_grid, n_folds = 3,
+                              scorer = NULL, n_cpu = 1){
+  
+  ### add default value to parameters
+  if(is.null(param_grid[["max_depth"]])) param_grid$max_depth <- Inf
+  if(is.null(param_grid[["margin"]])) param_grid$margin <- 0.0
+  if(is.null(param_grid[["min_sample"]])) param_grid$min_sample <- 0.0
+  if(is.null(param_grid[["loss"]])) param_grid$loss <- "hinge"
+  if(is.null(param_grid[["n_trees"]])) param_grid$ntrees <- 10
+  if(is.null(param_grid[["n_features"]])) param_grid$n_features <- c(as.integer(ncol(feature.mat)**0.5))
+  
+  ### check for unwanted parameters
+  assert_that(length(param_grid) <= 6, msg = "unexpected parameters as argument")
+  
+  ### combinations of all parameters grid
+  parameters <- expand.grid(max_depth = param_grid$max_depth, margin = param_grid$margin, 
+                            min_sample = param_grid$min_sample, loss = param_grid$loss,
+                            n_trees = param_grid$n_trees, n_features = param_grid$n_features)
+  parameters <- as.data.frame(parameters)
+  
+  ### for param grid run loop to check best value
+  cv_results <- NULL
+  best_result <- NULL
+  best_result$best_score <- attr(scorer, "worst")
+  
+  ### parallelize using foreach, see all permutation combination of param grid values
+  ### register parallel backend
+  if(n_cpu == -1) n_cpu <- detectCores() 
+  assert_that(detectCores() >= n_cpu)
+  
+  cl <- makeCluster(n_cpu)
+  registerDoParallel(cl)
+  
+  fitscore_result <- list()
+  fitscore_result <- foreach(i = 1:nrow(parameters), 
+                             .packages = "mmit") %dopar% 
+    fit_and_score(target.mat = target.mat, feature.mat = feature.mat, 
+                  parameters = parameters[i,], learner = "mmif", predict = mmif.predict,
```

Comment:
Auto build the name of the pred function by adding .predict to the learner

Reply:
done